docs: edit property binding doc (#38799)
angular/angular: aio/content/guide/property-binding.md

# Property binding

Property binding in Angular helps you set values for properties of HTML elements or directives.
With property binding, you can do things such as toggle button functionality, set paths programatically, and share values between components.

<div class="alert is-helpful">

See the <live-example></live-example> for a working example containing the code snippets in this guide.

</div>

## Prerequisites

To get the most out of property binding, you should be familiar with the following:

* [Basics of components](guide/architecture-components)
* [Basics of templates](guide/glossary#template)
* [Binding syntax](guide/binding-syntax)

<hr />

> ## Direction of data flow

Reviewer: ```suggestion
## Understanding the flow of data
```

Property binding moves a value in one direction, from a component's property into a target element property.

> For more information on listening for events, see [Event binding](guide/event-binding).

Reviewer: ```suggestion
<div class="alert is-helpful">

For more information on listening for events, see [Event binding](guide/event-binding).

</div>
```

To read a target element property or call one of its methods, see the API reference for [ViewChild](api/core/ViewChild) and [ContentChild](api/core/ContentChild).

> ## Property binding syntax `[property]`

Reviewer: ```suggestion
## Binding to a property
```

> To bind to an element's property, enclose it in square brackets, `[]`, which identifies the property as a target property.
> A target property is the DOM property to which you want to assign a value.
> The target property in the following code is the image element's `src` property.

Reviewer: ```suggestion
To bind to an element's property, enclose it in square brackets, `[]`, which identifies the property as a target property.
A target property is the DOM property to which you want to assign a value.
For example, the target property in the following code is the image element's `src` property.
```

To use the target and the property, you must declare them in their respective classes.

Declare the target of `childItem` in its component class, in this case `ItemDetailComponent`.

For example, the following code declares the target of `childItem` in its component class, in this case `ItemDetailComponent`.

Then, the code contains an `@Input()` decorator with the `childItem` property so data can flow into it.

<code-example path="property-binding/src/app/item-detail/item-detail.component.ts" region="input-type" header="src/app/item-detail/item-detail.component.ts"></code-example>

Next, the code declares the property of `parentItem` in its component class, in this case `AppComponent`.
In this example the type of `childItem` is `string`, so `parentItem` needs to be a string.
Here, `parentItem` has the string value of `lamp`.

<code-example path="property-binding/src/app/app.component.ts" region="parent-data-type" header="src/app/app.component.ts"></code-example>

With this configuration, the view of `<app-item-detaill>` uses the value of `lamp` for `childItem`.

> ## How property binding helps keep content secure

Reviewer: ```suggestion
## Property binding and security
```

> Consider the following malicious content.

Reviewer: ```suggestion
Property binding can help keep content secure. For example, consider the following malicious content.
```

<code-example path="property-binding/src/app/app.component.html" region="malicious-interpolated" header="src/app/app.component.html"></code-example>

The browser doesn't process the HTML and instead displays it raw, as follows.

<code-example language="bash">
"Template &lt;script&gt;alert("evil never sleeps")&lt;/script&gt; Syntax" is the interpolated evil title.
</code-example>

Angular does not allow HTML with `<script>` tags, neither with [interpolation](guide/interpolation) nor property binding, which prevents the JavaScript from running.

In the following example, however, Angular [sanitizes](guide/security#sanitization-and-security-contexts) the values before displaying them.

<code-example path="property-binding/src/app/app.component.html" region="malicious-content" header="src/app/app.component.html"></code-example>

Interpolation handles the `<script>` tags differently than property binding, but both approaches render the content harmlessly.
The following is the browser output of the sanitized `evilTitle` example.

<code-example language="bash">
"Template Syntax" is the property bound evil title.
</code-example>

> ## Property binding vs. interpolation

Reviewer: ```suggestion
## Property binding and interpolation
```

# Property binding

Property binding in Angular helps you set values for properties of HTML elements or directives.
With property binding, you can do things such as toggle button functionality, set paths programatically, and share values between components.

<div class="alert is-helpful">

See the <live-example></live-example> for a working example containing the code snippets in this guide.

</div>

## Prerequisites

To get the most out of property binding, you should be familiar with the following:

* [Basics of components](guide/architecture-components)
* [Basics of templates](guide/glossary#template)
* [Binding syntax](guide/binding-syntax)

<hr />

## Direction of data flow

> Property binding flows a value in one direction, from a component's property into a target element property.

Author: Hm, how about `moves`? That keeps it in the active. We do use the word flow in different parts of the documentation because of the concept of data flow, but I can go either way.

Here are some places we use it:
https://angular.io/guide/glossary#unidirectional-data-flow
https://angular.io/guide/glossary#input
https://angular.io/guide/inputs-outputs
https://angular.io/tutorial/toh-pt1#edit-the-hero

For more information on listening for events, see [Event binding](guide/event-binding).

To read a target element property or call one of its methods, see the API reference for [ViewChild](api/core/ViewChild) and [ContentChild](api/core/ContentChild).

## Property binding syntax `[property]`

To bind to an element's property, enclose it in square brackets, `[]`, which identifies it as the target property.
A target property is the thing to which you want to assign a value.
The target property in the following code is the image element's `src` property.

<code-example path="property-binding/src/app/app.component.html" region="property-binding" header="src/app/app.component.html"></code-example>

There's also the alternative `bind-` prefix :

<code-example path="property-binding/src/app/app.component.html" region="bind-prefix" header="src/app/app.component.html"></code-example>

> 

Author: Tbh, I don't know if the square brackets are preferable or not as I haven't seen any mention of a preference either way. I believe that the `bind-` syntax is the canonical form and that `[]` is shorthand, but I could be mistaken. Maybe an engineer could shed some light on this for us. I'll leave this conversation unresolved because I'd like to know more about it.

> 

Reviewer 2: The square braces are the preferred form. Ideally, I'd omit all reference to `bind-` (and it's equivalents for events and template variables) from any sort of introductory or conceptual guide, but I _would_ include it in a more comprehensive guide. The historical context was that some people found Angular's `[property]` and `(event)` syntax off-putting, so this was introduced to provide a more traditional alternative. I have personally never _seen_ an application using the alternative, but I'm sure there are some out there. I wouldn't want this to become a secret, undocumented feature.

I added an item to the next framework meeting to discuss whether we actually just want to go ahead and deprecate this syntax.

> 

Reviewer: @kapunahelewong I suggest removing this. We could sock it away somewhere in case we need to add documentation later.

I think it's important our documentation focus on "the preferred way to work". Let me know if you want to discuss this further though!

> 

Reviewer: Upon reflection, I think removing the `bind-prefix` content is the right thing to do.

The value of `classes` is `special`.

<code-example path="property-binding/src/app/app.component.ts" region="directive-property" header="src/app/app.component.ts"></code-example>

In this way, Angular applies the class `special` to the `<p>` element so that you can use `special` to apply CSS styles.

## Bind values between components

To set the model property of a custom component, place the target, here `childItem`, between square brackets `[]` followed by an equal sign and the property.
Here, the property is `parentItem`.

<code-example path="property-binding/src/app/app.component.html" region="model-property-binding" header="src/app/app.component.html"></code-example>

To use the target and the property, you must declare them in their respective classes.

Declare the target of `childItem` in its component class, in this case `ItemDetailComponent`.
Use the `@Input()` decorator with the `childItem` property so data can flow into it.

<code-example path="property-binding/src/app/item-detail/item-detail.component.ts" region="input-type" header="src/app/item-detail/item-detail.component.ts"></code-example>

> Declare the property of `parentItem` in its component class, in this case `AppComponent`.

Author: Since this is a task in a series of steps in the imperative, it seems clearer to tell the reader what the next thing is they need to do, that is declaring the property. So the person declares the property as the agent/subject rather than the code doing so. The code declaring it could make sense if we were giving someone a tour of a demo as opposed to instructing them on the steps of the task. If we switch this statement from the imperative, then we should switch the other steps in this task, and then maybe convert it to a concept. Do you have a preference here of concept versus task?

In this example the type of `childItem` is `string`, so `parentItem` needs to be a string.
Here, `parentItem` has the string value of `lamp`.

<code-example path="property-binding/src/app/app.component.ts" region="parent-data-type" header="src/app/app.component.ts"></code-example>

With this configuration, the view of `<app-item-detaill>` uses the value of `lamp` for `childItem`.

> ## How property binding keeps content secure

Reviewer: ```suggestion
## How property binding helps keep content secure
```
We'll probably see a lot of issues like this, but it's really important we don't make promises we can't keep. So something doesn't *make* something secure. It can, however, *help* make something secure. That's an important distinction.

> Imagine the following malicious content.

Reviewer: ```suggestion
Consider the following malicious content.
```

<code-example path="property-binding/src/app/app.component.ts" region="malicious-content" header="src/app/app.component.ts"></code-example>

The component template interpolates the content as follows:

<code-example path="property-binding/src/app/app.component.html" region="malicious-interpolated" header="src/app/app.component.html"></code-example>

The browser doesn't process the HTML and instead displays it raw, as follows.

<code-example language="bash">
"Template &lt;script&gt;alert("evil never sleeps")&lt;/script&gt; Syntax" is the interpolated evil title.
</code-example>

Angular does not allow HTML with `<script>` tags, neither with [interpolation](guide/interpolation) nor property binding, which prevents the JavaScript from running.

In the following example, however, Angular [sanitizes](guide/security#sanitization-and-security-contexts) the values before displaying them.

<code-example path="property-binding/src/app/app.component.html" region="malicious-content" header="src/app/app.component.html"></code-example>

> Interpolation handles the `<script>` tags differently than property binding but both approaches render the content harmlessly.

Reviewer: ```suggestion
Interpolation handles the `<script>` tags differently than property binding, but both approaches render the content harmlessly.
```

The following is the browser output of the sanitized `evilTitle` example.

<code-example language="bash">
"Template Syntax" is the property bound evil title.
</code-example>

## Property binding vs. interpolation

Often [interpolation](guide/interpolation) and property binding can achieve the same results.
The following binding pairs do the same thing.

<code-example path="property-binding/src/app/app.component.html" region="property-binding-interpolation" header="src/app/app.component.html"></code-example>

> When rendering data values as strings, either form is appropriate, though readability tends to favor interpolation.

Reviewer: ```suggestion
When rendering data values as strings, it is recommended that you use interpolation for code readability, although either form is appropriate.
```

Property binding moves a value in one direction, from a component's property into a target element property.

For more information on listening for events, see [Event binding](guide/event-binding).

To read a target element property or call one of its methods, see the API reference for [ViewChild](api/core/ViewChild) and [ContentChild](api/core/ContentChild).

## Property binding syntax `[property]`

To bind to an element's property, enclose it in square brackets, `[]`, which identifies the property as a target property.
A target property is the DOM property to which you want to assign a value.
The target property in the following code is the image element's `src` property.

<code-example path="property-binding/src/app/app.component.html" region="property-binding" header="src/app/app.component.html"></code-example>

There's also the alternative `bind-` prefix :

<code-example path="property-binding/src/app/app.component.html" region="bind-prefix" header="src/app/app.component.html"></code-example>

In most cases, the target name is the name of a property, even when it appears to be the name of an attribute.
In this example, `src` is the name of the `<img>` element property.

> The brackets, `[]`, cause Angular to evaluate the template expression, which is to the right of the equal sign.
> If you omit the brackets, Angular treats the string as a constant and initializes the target property with that string:

Reviewer 2: ```suggestion
The brackets, `[]`, cause Angular to evaluate the right-hand-side of the assignment as a dynamic expression.
Without the brackets, Angular treats the the right-hand-side as a string literal and sets the property to that static value:
```
* I think using "right-hand-side" and "assignment" here help clarify what we're talking about (not 100% sure about hyphenation rules)
* I added the word "dynamic" in front of "template expression" to emphasize the difference between the case without brackets
* "String literal" is a bit more correct than "constant", since the initial value of a constant (in a genral sense) can still be determined via some expression/evaluation.

<code-example path="property-binding/src/app/app.component.html" region="no-evaluation" header="src/app.component.html"></code-example>

Omitting the brackets renders the string `parentItem`, not the value of `parentItem`.

> ### When to omit the brackets

Reviewer 2: This might be too philosophical, but I wouldn't frame this as "When to omit the braces" because not having braces is just plain old vanilla HTML (which I believe we expect people to know already). Instead, I would focus on the concept of using bindings to _augment_ the view with dynamic values that Angular automatically keeps up to date (this is Angular's special value addition).

Author: Do you think we need to remove this whole section? The section focuses on one-time initialization, but like you pointed out, that would just be regular HTML (so maybe we don't need it?). Here's what I have tentatively updated it to, but I'd like to make sure it's what you intend. I also changed the next couple of sentences. What do you think?

### One-time string initialization
```suggestion
### One-time string initialization

Bindings augment the view with dynamic values that Angular automatically keeps up-to-date.
However, if the value doesn't need to change, you can set the string one time.

For one-time string initialization, omit the square brackets, `[]`, when all of the following are true:

```

Reviewer 2: I might lean on the side of just omitting this section entirely and cueing this idea as part of the other content. If I were explaining this concept to someone new, I would start off with a completely static html template, and then make the content dynamic by adding bindings. That might be too tutorial-y, though. @aikidave thoughts?

Reviewer: I am all for deleting content. Let's do it!

Don't use the square brackets, `[]`, in cases of one-time string initialization when all of the following are true:

* The target property accepts a string value.
* The string is a fixed value that you can put directly into the template.
* This initial value never changes.

As with standard HTML, one-time string initialization also works for directive and component property initialization.
The following example initializes the `prefix` property of the `StringInitComponent` to a fixed string, not a template expression.
Angular sets the value only once.

<code-example path="property-binding/src/app/app.component.html" region="string-init" header="src/app/app.component.html"></code-example>

> <!-- To do: is this content valid?

Reviewer 2: For this section, the content is accurate, but personally I see this as another area that highlights the need for a separate "main concepts" and "comprehensive guide"- most of the time, you don't need to know or think about the distinction between a native element property and a directive/component input. The main area where you might need to know is for a UI component library, and if the library is well-written, users shouldn't need to know the difference in the majority of cases.

Author: Thank you so much, Jeremy! This is super helpful. 
cc @aikidave

Reviewer: I'm with Jeremy. Let's find a place to archive this information and decide later how important it is. Otherwise, I vote we err on making the content easier to understand/read.

Author: I've moved this information into a new document called `binding-concepts.md` along with the above section about one time string initialization. Is this what you're thinking, @aikidave?

Reviewer: Upon reflection, I don't think we need an additional topic in the Conceptual Reference section. Instead, let's take that content and make it more task-oriented. So `One-time string initialization` becomes `Setting a value once` (or similar), and `Property binding and directives` becomes `Distinguishing between properties and directives` (or similar). What do you think?

</code-example>

Angular does not allow HTML with `<script>` tags, neither with [interpolation](guide/interpolation) nor property binding, which prevents the JavaScript from running.

In the following example, however, Angular [sanitizes](guide/security#sanitization-and-security-contexts) the values before displaying them.

<code-example path="property-binding/src/app/app.component.html" region="malicious-content" header="src/app/app.component.html"></code-example>

Interpolation handles the `<script>` tags differently than property binding, but both approaches render the content harmlessly.
The following is the browser output of the sanitized `evilTitle` example.

<code-example language="bash">
"Template Syntax" is the property bound evil title.
</code-example>

## Property binding vs. interpolation

Often [interpolation](guide/interpolation) and property binding can achieve the same results.
The following binding pairs do the same thing.

<code-example path="property-binding/src/app/app.component.html" region="property-binding-interpolation" header="src/app/app.component.html"></code-example>

> When rendering data values as strings,  it is recommended that you use interpolation for code readability, although either form is appropriate.

Reviewer 2: Extra space, and also passive voice (it is recommended...)
